### DISCUSSION TITLE
Global Context type

### DISCUSSION THREAD

**Discussion by bombillazo** [Urigo/graphql-modules#1853] [Q&A] [answered]
Hello, I am trying to follow [these instructions](https://www.graphql-modules.com/docs/essentials/type-safety#shaping-context-type) on typing the Context but don't know where I need to declare that Global namespace.

**Comment by ardatan** [comment]
You can define it anywhere. It should work

**Reply by bombillazo** [reply]
Thanks for answering. I am new to using namespaces, I added that expression to my common module but I still see the context type as `any` .

I am using GraphQL Express server and want to add the Express Req to my Context.

**Reply by bombillazo** [reply]
@ardatan Would you happen to have an example in a repo?

**Comment by bombillazo** [comment]
Don't know what else is required so my context gets Globally typed. 

![image](https://user-images.githubusercontent.com/44002321/133269187-b02d058e-411b-4088-86fd-7893b10b7ef5.png)

**Comment by bombillazo** [comment, marked as answer]
Ok got it to work the following manner:

1. Add the declare statement in your code (as specified above). I have mine in the `index.ts` file where I create my graphql-modules application.

```ts
declare global {
  namespace GraphQLModules {
    interface GlobalContext {
      myProp: string; 
    }
  }
}  

const application = createApplication({
...
```
2. Use GraphQL Code Generator [typescript-resolvers](https://www.graphql-code-generator.com/docs/plugins/typescript-resolvers) plugin to add the context type to my resolvers. I am doing it programmatically but it's the same concept:

```yml
generates:
path/to/file.ts:
  plugins: 
    - typescript
    - typescript-resolvers
  config:
    contextType: "GraphQLModules.Context"
...
```

After that my resolver context gets typed! :D 

![image](https://user-images.githubusercontent.com/44002321/133327466-7a2fda9a-afc2-4abd-b539-c6c5161ac279.png)

**Reply by ardatan** [reply]
Also see this one;
https://www.graphql-code-generator.com/docs/presets/graphql-modules

**Reply by bombillazo** [reply]
I was already using that preset, still had to add the `contextType` to the plugin:

```ts
await generate(
    {
      require: [],
      schema: [
        './src/modules/**/typeDefs.ts',
      ],
      generates: {
        './src/modules/': { 
          preset: 'graphql-modules',
          config:{
            contextType: 'GraphQLModules.Context',
          },
          presetConfig: {
            baseTypesPath: '../@generated/schemaTypes.ts',
            filename: 'generatedTypes.ts',
          },
          plugins: [
            'typescript',
            'typescript-resolvers'
          ],
        },
      },
    },
    true
  );
```

**Reply by Urigo** [reply]
@bombillazo is there something we could add to the docs to make it easier for others?

**Reply by bombillazo** [reply]
Hey @Urigo  , I think what confused me was understanding that modifying the `GraphQLModules.GlobalContext` using namespace automatically would reflect the type in the context input. I still needed to set the type manually in my resolvers, or (how I did it) use Code Generator to set my Context type for all resolvers. 

I'm a beginner at TypeScript and GraphQL so it wasn't clear to me. Explicitly mentioning that you still need to set the type somewhere would've made it clearer for me. I thought Modules did that for me.

**Reply by Urigo** [reply]
Thanks for sharing @bombillazo !

Do you think you are able to create a draft pull request with the wording that would be best for you?
https://github.com/Urigo/graphql-modules/edit/master/website/docs/essentials/type-safety.md

I believe you would be best to structure it as you are the one who was in this position just now.

I would love to help of course

**Reply by bombillazo** [reply]
Hey @Urigo , a PR I mas was merged, but I dont see the docs have been updated

**Reply by Urigo** [reply]
it should be updated now

**Reply by enheit** [reply]
@Urigo Hey, I just read your tread and want to say that I also had an issue with context type. Would be nice if you explicitly show where to put this definition. 

Firstly, I tried to put this code inside `graphql-context.d.ts` file and add it to the [`typeRoots`](https://www.typescriptlang.org/tsconfig#typeRoots) of `tsconfig.json` file of my project. It didn't help

After investigating @bombillazo code I decided to put it beside `createApplication({ ... })` call. It works, but show warnings

> ES2015 module syntax is preferred over custom TypeScript modules and namespaces. (@typescript-eslint/no-namespace)ts-standard(@typescript-eslint/no-namespace)

So, would be nice to have a little more information on that in your documentation
